tripal/tripal
# Implement the fallback entity lookup

# New Feature

### Closes #1850 

### Tripal Version: 4.x

## Description
There are four cases in core Tripal where links use a different name, currently entity links do not work for these.
The goofy fields that we need to test are listed at the end of this comment https://github.com/tripal/tripal_doc/issues/32#issuecomment-1837532399
```
arraydesign FOREIGN KEY (manufacturer_id) REFERENCES contact(contact_id)
assay FOREIGN KEY (operator_id) REFERENCES contact(contact_id)
biomaterial FOREIGN KEY (biosourceprovider_id) REFERENCES contact(contact_id)
biomaterial FOREIGN KEY (taxon_id) REFERENCES organism(organism_id)
```

We had originally intended to consult the third party settings, but for reasons of efficiency I decided to have the TripalTypes property pass through the table name. It already knows this, and thus avoids a chunk of code to get at the third party settings to look it up.

Because I changed the entity_id property, we need an update function. Again for efficiency, I recycled the existing update function, so as not to duplicate it, it will just be run again with a higher number.

## Testing?
 1. Create your docker on the 4.x branch so we can test the update function
 2. Change to this branch `git checkout tv4g1-issue1850-entity-lookup-third-party-settings`
 3. Run the update `docker updatedb` - verify no errors are reported
 4. Import the gene expression type collection
 5. Create an organism
 6. Create a contact
 7. Create an array design. You will need a contact that you created in step 6
 8. Verify that the "Manufacturer" is a link 
 9. Create an assay. Use the contact as the "Operator"
 10. Again verify that the "Operator" is a link
 11. Create a biomaterial.
 12. Verify that both biosource provider and Organism (taxon id) are links
 13. I have added a kernel test for the fallback method (manufacturer on arraydesign) so make sure tests pass 🙌
